libraries/Microsoft.Bot.Builder.Ai/TranslationMiddleware.cs

```diff
@@ -0,0 +1,177 @@
+                    {
+                        languageChanged = await _setUserLanguage(context);
+                    }
+
+                    if (!languageChanged)
+                    {
+                        // determine the language we are using for this conversation
+                        var sourceLanguage = "";
+                        if (_getUserLanguage == null)
+                            sourceLanguage = await Task.Run(() => _translator.Detect(message.Text)); // context.Conversation.Data["Language"]?.ToString() ?? this._nativeLanguages.FirstOrDefault() ?? "en";
+                        else
+                        {
+                            sourceLanguage = _getUserLanguage(context);
+                        } 
+                        if (_patterns.ContainsKey(sourceLanguage) && _patterns[sourceLanguage].Count>0)
+                        {
+                            this._translator.SetPostProcessorTemplate(_patterns[sourceLanguage]);
+                        }
+                        if (!_nativeLanguages.Contains(sourceLanguage))
+                        {
+                            var translationContext = new TranslationContext
+                            {
+                                SourceText = message.Text,
+                                SourceLanguage = sourceLanguage,
+                                TargetLanguage = (this._nativeLanguages.Contains(sourceLanguage)) ? sourceLanguage : this._nativeLanguages.FirstOrDefault() ?? "en"
+                            };
+
+                            // translate to bots language
+                            if (translationContext.SourceLanguage != translationContext.TargetLanguage)
+                                await TranslateMessageAsync(context, message, translationContext.SourceLanguage, translationContext.TargetLanguage).ConfigureAwait(false);
+                        }
+                    }
+                }
+            }
+            await next().ConfigureAwait(false);
+        }
+
+
+        private static readonly Regex UrlRegex = new Regex(@"(https?://[^\s]*)", RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.Multiline);
+
+        /// <summary>
+        /// Translate .Text field of a message
+        /// </summary>
+        /// <param name="context"/>
+        /// <param name="message"></param>
+        /// <param name="sourceLanguage"/>
+        /// <param name="targetLanguage"></param>
+        /// <returns></returns>
+        private async Task TranslateMessageAsync(ITurnContext context, IMessageActivity message, string sourceLanguage, string targetLanguage)
+        {
+            // if we have text and a target language
+            if (!String.IsNullOrWhiteSpace(message.Text) && !String.IsNullOrEmpty(targetLanguage))
+            {
+                if (targetLanguage == sourceLanguage)
+                    return;
+
+                var text = message.Text.Length <= 65536 ? message.Text : message.Text.Substring(0, 65536);
```

Comment:
Magic numbers in code are bad. Please make constants out of these, link to where them come from, explain them, and give them good names. 

For example, is the rule here: 64Kb of data? 65536 UTF-16 codepoints? 65535 Graphemes? 
As this is written, each character is 16-bits long, meaning 65535 characters is actual 128kb of data.  Because I don't know what you're testing for, I can't tell if this is right or not.

Reply:
this test is from the initial translator middleware class , i think we don't need it